clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -85,6 +90,10 @@
     private int drainIndex;
     private final TransactionManager transactionManager;
 
+    // A mapping of batch's creation time to the batch for quick of the oldest batch
+    // and lookup in the order of creation
+    private final NavigableMap<Long, List<ProducerBatch>> inFlightBatches;
```

Comment:
Apparently the my understanding of `TreeSet` is not accurate. It uses the comparator to decide whether the entries are the same or not. We can use a TreeMap<Long, Set<ProducerBatch>> then. We may also want to bucket the timestamp a little bit, say one second to avoid huge amount of Sets created for each ms in the `TreeMap`.

Reply:
I was thinking about this too. Using millisecond as unit for Map key is not prudent.

After the switch to second as unit, we may need to check the two adjacent buckets keyed by ts-1  (sec) and ts+1 (sec).

Reply:
Ok so the naming is consistent. A `TreeSet` is a set it's just that equality criterion is different.